[cybelangel] New external-import connector for the CybelAngel platform

Lhorus6 opened the pull request:
### Proposed changes

* Creation of CybelAngel connector

### Related issues

* https://github.com/OpenCTI-Platform/connectors/issues/4450

### Checklist

- [X] I consider the submitted work as finished
- [X] I have signed my commits using GPG key.
- [X] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [X] Where necessary I refactored code to improve the overall quality

SamuelHassine committed:
[cybelangel] New external-import connector for the CybelAngel platform

Imports CybelAngel claimed-attack threat intelligence into OpenCTI as
STIX 2.1 objects. For every claim returned by
``/api/v1/threat-intelligence/claimed-attacks`` the connector creates
Intrusion Sets for the threat actors, Locations for the targeted
countries, Identity objects for the victim organizations and impacted
industries, a Campaign per (threat actor, victim) pair, and the
corresponding ``targets`` / ``attributed-to`` relationships.

Implementation notes (addressing the inline review):

- TLP handling is now table-driven: ``TLP:WHITE`` / ``TLP:CLEAR`` reuse
  the standard ``stix2.TLP_WHITE``; ``TLP:GREEN`` / ``TLP:AMBER`` /
  ``TLP:RED`` reuse the standard objects (no more invalid
  ``ExtraPropertiesError``); only ``TLP:AMBER+STRICT`` is created as a
  custom statement marking.
- ``last_run`` is stored and read as UTC ISO 8601, so the
  ``start_date`` / ``end_date`` query parameters are always
  time-zone-consistent.
- The API URL is aligned across code / sample / docker-compose /
  README / docs to ``https://platform.cybelangel.com`` and the endpoint
  is ``/api/v1/threat-intelligence/claimed-attacks`` with ``skip`` /
  ``limit`` / ``start_date`` / ``end_date`` parameters, matching the
  official documentation.
- Scheduling is delegated to ``OpenCTIConnectorHelper.schedule_iso``
  (ISO 8601 ``CONNECTOR_DURATION_PERIOD``, default ``PT6H``).
- HTTP requests now use explicit ``timeout=(10, 60)`` and the data
  fetch loop is wrapped in ``requests.RequestException`` handling.
- Authentication failures no longer log "Token received" — the success
  log is only emitted after a non-empty token is confirmed.
- ``IntrusionSet.resource_level`` now uses the lowercase STIX enum
  values (``contest`` / ``organization``).
- Pagination stops early when a claim older than the previous
  ``last_run`` is reached (results are sorted by ``claimed_at desc``).
- ``opencti_bundle`` returns a success flag; ``last_run`` is advanced
  and the work is marked processed only on success.
- ``work_id`` is initialised to ``None`` before the try block so an
  exception raised before ``initiate_work`` succeeds no longer raises
  ``UnboundLocalError``.
- ``send_stix2_bundle`` no longer passes the deprecated ``update``
  argument.
- Config loading uses ``with open(...)`` + ``yaml.safe_load``.

Other:
- Dockerfile collapsed to a single ``RUN`` (no more ``apk upgrade``,
  no more duplicate ``pip install`` layer).
- ``docker-compose.yml`` uses the ``rolling`` tag and a reachable
  ``http://opencti:8080`` URL.
- ``requirements.txt`` pins ``pycti`` to the current release and bounds
  ``stix2`` to ``>=3.0.1,<4``.
- README rewritten with single-pipe Markdown tables, the correct
  default URL, Python 3.12 requirement, an authentication note about
  the 1-hour token TTL / 2,000/month rate limit, and a behaviour
  section that matches the implementation.
- ``__metadata__/connector_manifest.json`` added.

Closes #4450

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+        }
+        for attempt in range(1, max_retries + 1):
+            self.helper.connector_logger.info(
+                f"Attempt {attempt} to authenticate with CybelAngel API"
+            )
+            try:
+                response = requests.post(
+                    self.cybelangel_auth_url, json=auth_data, headers=headers
+                )
```
Fixed: both the auth `POST` and the data `GET` now use an explicit `timeout=(10, 60)`. Network errors raised by `requests` are caught in `_fetch_and_process_pages` and retried up to 3 times with a back-off.

SamuelHassine commented on external-import/cybelangel/src/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+pycti==6.7.9
+requests>=2.31.0
+stix2>=3.0.1
```
Fixed: `requirements.txt` now pins `stix2>=3.0.1,<4` (plus a current `pycti` and bounded `requests` / `PyYAML`).

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+
+        # Resource level heuristic
+        resource_level = (
+            "Contest" if "ddos" in campaign_objective.lower() else "Organization"
```
Fixed: `resource_level` now uses the lowercase STIX 2.1 enum values (`contest` / `organization`).

SamuelHassine commented on external-import/cybelangel/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## Requirements
+
+- Python 3.10+
```
Fixed: README now requires Python >= 3.12, matching the `python:3.12-alpine` Dockerfile.

SamuelHassine commented on external-import/cybelangel/README.md:
```diff
@@ -0,0 +1,47 @@
+
+| Parameter                   | Docker env var              | Mandatory | Description                                                                                              |
+|-----------------------------|-----------------------------|-----------|----------------------------------------------------------------------------------------------------------|
+| `opencti_url`               | `OPENCTI_URL`               | Yes       | The URL of the OpenCTI platform.                                                                         |
```
Fixed: README rewritten with standard single-pipe Markdown tables.

SamuelHassine commented on external-import/cybelangel/README.md:
```diff
@@ -0,0 +1,47 @@
+
+## Development Notes
+
+- The `published_at_range` parameter is dynamically calculated using the current date and the `fetch_period`.
```
Fixed: the `published_at_range` mention is gone. The README now documents the actual `start_date` / `end_date` query parameters used by `/api/v1/threat-intelligence/claimed-attacks`.

SamuelHassine commented on external-import/cybelangel/docker-compose.yml:
```diff
@@ -0,0 +1,21 @@
+  connector-cybelangel:
+    image: opencti/connector-cybelangel:6.7.9
+    environment:
+      - OPENCTI_URL=http://localhost
```
Fixed: `docker-compose.yml` now uses `OPENCTI_URL=http://opencti:8080` (the typical service name inside a Compose network).

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+                yaml.load(open(config_file_path), Loader=yaml.FullLoader)
+                if os.path.isfile(config_file_path)
+                else {}
+            )
```
Fixed: config is now loaded with `with open(...) as config_file: yaml.safe_load(config_file)`.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+    # ----------------------
+    # Fetch Data
+    # ----------------------
+    def opencti_bundle(self, work_id, last_run=None):
+        token = self.authenticate()
+        self.helper.connector_logger.debug(
+            "Token received, proceeding with data fetching."
+        )
+        cybelangel_org_identity = self.create_cybelangel_org()
+
+        if not token or not cybelangel_org_identity:
+            self.helper.connector_logger.error("Missing token or CybelAngel identity.")
+            return
+
```
Fixed: `opencti_bundle` now early-exits when `authenticate()` returns `None`, and the success log is only emitted after a non-empty token is confirmed.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+            self.helper.connector_logger.debug(
+                f"Fetching data from CybelAngel API: {url}"
+            )
+            response = requests.get(url, headers=headers)
```
Fixed: `requests.get` is now wrapped in `try/except requests.RequestException` with the same 3-attempt back-off as the 5xx / non-200 retry branch, and uses an explicit `timeout=(10, 60)`.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+                f"Stopping processing as claimed_at {claimed_at.strftime('%Y-%m-%dT%H:%M:%SZ')} is before "
+                f"last_run value {since_date.strftime('%Y-%m-%dT%H:%M:%SZ')}"
+            )
+            return
```
Fixed: the pagination loop now stops entirely (`return True`) as soon as the first claim older than `since_date` is reached, instead of just returning from `_process_attack`. There's a new dedicated helper `_is_attack_too_old` that owns the comparison.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+            error_message = f"Unexpected error during synchronization: {str(e)}"
+            self.helper.api.work.report_expectation(
+                work_id=work_id,
+                error={"error": error_message, "source": "CybelAngel Connector"},
```
Fixed: `work_id` is now initialised to `None` before the `try` block, and the `except` branch only calls `report_expectation` / `to_processed(..., in_error=True)` when `work_id is not None`.

SamuelHassine commented on external-import/cybelangel/src/config.yml.sample:
```diff
@@ -0,0 +1,20 @@
+  auth_url: 'https://auth.cybelangel.com/oauth/token'
+  marking: 'TLP:AMBER+STRICT'
+  fetch_period: 7 # In days
+  duration_period: "PT6H" # every 6 hours
```
Fixed: `duration_period` is now under the `connector:` section in `config.yml.sample` (and `CONNECTOR_DURATION_PERIOD` in `docker-compose.yml`), matching what the code reads via `get_config_variable('CONNECTOR_DURATION_PERIOD', ['connector', 'duration_period'], ...)`.

SamuelHassine commented on external-import/cybelangel/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    pip install --no-cache-dir -r requirements.txt && \
+    apk del git build-base libxslt-dev libxml2-dev
+
+
```
Fixed: the Dockerfile now installs dependencies in a single `RUN` block (no more duplicate `pip install -r requirements.txt`).

SamuelHassine commented on external-import/cybelangel/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+    pip install --no-cache-dir -r requirements.txt && \
+    apk del git build-base libxslt-dev libxml2-dev && \
+    rm -rf /var/cache/apk/*
+
```
Fixed: `apk update && apk upgrade` is gone; the Dockerfile now uses `apk --no-cache add ...` like the other connectors.

SamuelHassine commented on external-import/cybelangel/README.md:
```diff
@@ -0,0 +1,47 @@
+| `connector_log_level`       | `CONNECTOR_LOG_LEVEL`       | No        | Log output for the connector. Defaults to `error`.                                                       |
+| `cybelangel_client_id`      | `CYBELANGEL_CLIENT_ID`      | Yes       | The client ID provided by CybelAngel.                                                                    |
+| `cybelangel_client_secret`  | `CYBELANGEL_CLIENT_SECRET`  | Yes       | The client secret provided by CybelAngel.                                                                |
+| `cybelangel_api_url`        | `CYBELANGEL_API_URL`        | No        | Defaults to `https://api.cybelangel.com/v1`.                                                             |
```
Fixed: the README, the code default, `docker-compose.yml` and `config.yml.sample` are now all aligned on `https://platform.cybelangel.com` (the value used by the official documentation).

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,799 @@
+            self.helper.set_state({"last_run": now.astimezone().isoformat()})
+            message = "End of synchronization"
+            self.helper.api.work.to_processed(work_id, message)
+            self.helper.connector_logger.info(message)
```
Fixed: `opencti_bundle` / `_fetch_and_process_pages` now return a `bool` success flag. `process_data` only advances `last_run` and calls `to_processed(..., in_error=False)` when the flag is `True`; otherwise the work is marked `in_error` and the cursor is left untouched so the next run retries from the same window.

SamuelHassine commented:
@Megafredo The PR has been completely reworked on top of current master and squashed into a single GPG-signed commit. Every blocking point (API URL alignment, `load_marking_definition`, `rolling` docker tag, `last_run` UTC consistency) and every Copilot thread above have been addressed and resolved. The endpoint / pagination / field names now match the official documentation (`/api/v1/threat-intelligence/claimed-attacks`, `skip`/`limit`, `start_date`/`end_date`, `claimed_at`, `countries`, `industries`, `victims`, `domains`). README documents the 1-hour token TTL / 2,000-tokens/month rate limit and the connector now uses `schedule_iso` for auto-backpressure. `send_stix2_bundle` no longer passes the deprecated `update=` argument, and the bundle / `last_run` / work state are only updated when the run completes successfully.

SamuelHassine reviewed (dismissed):
Full review done.

Rebased the entire change set onto current master and squashed everything into a single GPG-signed commit. All 17 Copilot review threads and every blocking point raised by @Megafredo have been addressed and resolved:

Code:
- TLP handling is now table-driven; `TLP:GREEN` / `TLP:AMBER` / `TLP:RED` reuse the standard stix2 markings (no more `ExtraPropertiesError`), `TLP:CLEAR` is mapped to `stix2.TLP_WHITE`, and only `TLP:AMBER+STRICT` is constructed as a custom statement marking.
- API endpoint, pagination params and response fields are aligned with the [official docs](https://developers.cybelangel.com/docs/cybelangel-platform-api/74e0222802605-get-claimed-attacks): `/api/v1/threat-intelligence/claimed-attacks`, `skip`/`limit`, `start_date`/`end_date`, `claimed_at`, `countries`, `industries`, `victims`, `domains`.
- `last_run` is read / stored as UTC ISO 8601, so the date-range parameter is always time-zone-consistent.
- `IntrusionSet.resource_level` now uses lowercase STIX enum values (`contest` / `organization`).
- HTTP requests use explicit `timeout=(10, 60)` and the data fetch loop catches `requests.RequestException` with the same 3-attempt back-off as the non-200 retry branch.
- Authentication failures no longer log "Token received" before the token is confirmed.
- Pagination stops entirely (not just per-attack) once a claim older than `last_run` is reached - results are sorted by `claimed_at desc`.
- `work_id` is initialised to `None` before the `try` block; the `except` branch only touches the work when it actually exists.
- `opencti_bundle` returns a success flag; `last_run` is advanced and the work is marked processed only on success. On failure the work is reported `in_error` and the cursor is left untouched so the next run retries from the same window.
- `send_stix2_bundle` no longer passes the deprecated `update=` argument.
- Config loading uses `with open(...) as f: yaml.safe_load(f)`.

Infrastructure:
- Dockerfile collapsed to a single `RUN` (no more `apk upgrade`, no more duplicate `pip install` layer).
- `docker-compose.yml` uses the `rolling` tag and a reachable `http://opencti:8080` URL.
- `requirements.txt` pins `pycti` to the current release and bounds `stix2` to `>=3.0.1,<4`.
- `config.yml.sample` moves `duration_period` under `connector:` to match the code.

Docs:
- README rewritten with single-pipe Markdown tables, the correct default URL, Python 3.12 requirement, an authentication note about the 1-hour token TTL / 2,000-tokens/month rate limit, and a behaviour section that matches the implementation. The misleading `published_at_range` mention is gone.
- `__metadata__/connector_manifest.json` added.
- PR title realigned with the master `[<slug>] <description>` convention.

CI is fully green (`PR conventions checks`, `Test tests/test-requirements.txt`, `Baseline coverage`, `Do Not Merge label check`, `codecov/patch`, `codecov/project`, `filigran/cla`). Locally: `isort==6.0.1 --profile black` / `black==25.1.0 --check` / `flake8 --ignore=E,W` clean, and `pylint --enable=no_generated_id_stix,no-value-for-parameter,unused-import` with the `linter_stix_id_generator` plugin → 10/10.

SamuelHassine committed:
[cybelangel] Address remaining review comments

* `cybelangel.py`: align `Identity.generate_id` with the STIX
  ``identity_class`` value. The CybelAngel author and the victim
  identities now use ``"organization"`` (lowercase, canonical STIX
  2.1 value) for both `generate_id` and `identity_class`, so the
  deterministic IDs stay consistent across runs and across other
  connectors that follow the same convention.
* `cybelangel.py`: claims with a missing or unparseable
  ``claimed_at`` are now skipped with a warning instead of being
  fed into ``Campaign.created`` / ``first_seen`` / ``last_seen`` /
  relationship ``created`` (which would raise during STIX object
  creation and abort the whole run). The dead ``if claimed_at:``
  branch around ``campaign_date`` is removed, and the per-claim
  parsing loop is the same as ``_is_attack_too_old`` so behavior
  is consistent.
* `cybelangel.py`: `_build_fetch_parameters` no longer crashes when
  the stored ``last_run`` is corrupted *and* ``CYBELANGEL_FETCH_PERIOD``
  is ``"all"``. The fallback path is now factored out into
  `_fetch_parameters_from_period`, which understands ``"all"``,
  numeric values (``int`` / ``str``) and invalid values (warns and
  falls back to full history). The docstring is updated to match
  the actual query string (``sort_by=claimed_at&sort_order=desc``).
* `entrypoint.sh`: ``python3 cybelangel.py`` is now invoked with
  ``exec`` (and the script enables ``set -e``) so Docker can
  forward ``SIGTERM`` / ``SIGINT`` directly to the Python process
  and perform a graceful shutdown.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,887 @@
+
+        """
+        try:
+            identity = stix2.Identity(
+                id=Identity.generate_id("CybelAngel", "Organization"),
+                spec_version="2.1",
+                name="CybelAngel",
+                description="Cybelangel is a cybersecurity company that specializes in detecting and mitigating cyber "
+                "threats.",
+                confidence=50,
+                identity_class="organization",
+                type="identity",
```
Fixed: both `Identity.generate_id` calls now use the canonical lowercase STIX `identity_class` value (`organization`), so:

- the CybelAngel author identity uses `Identity.generate_id(CybelAngel, organization)` and `identity_class=organization` (was `Organization` for the id, `organization` for the field);
- the victim identities go through `_create_identity(v, organization, ...)`.

This keeps the deterministic IDs consistent across runs and aligned with other connectors that follow the lowercase convention. Industry / sector identities already used the lowercase `class` value.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,887 @@
+                )
+                self.helper.connector_logger.warning(
+                    f"Invalid last_run format. Using last {fetch_period} days."
+                )
```
Fixed: factored the period-handling code out of `_build_fetch_parameters` into a dedicated `_fetch_parameters_from_period` helper that both branches (no `last_run` + corrupted `last_run`) now reuse. It supports:

- `all` (case-insensitive) and empty â†’ no date filter, returns `(None, None, base_sort)`;
- numeric value (`int` / `str`) â†’ builds a `start_date` / `end_date` window of N days;
- anything else â†’ warns (`Invalid CYBELANGEL_FETCH_PERIOD value, falling back to full history`) and returns `(None, None, base_sort)`.

So `CYBELANGEL_FETCH_PERIOD=all` combined with a corrupted cursor no longer raises `ValueError` on `int(fetch_period)`.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,887 @@
+            "Authorization": f"Bearer {token}",
+            "Content-Type": "application/json",
+        }
+
+        return self._fetch_and_process_pages(
+            headers, parameters, cybelangel_org_identity, work_id, since_date
+        )
+
+    def _build_fetch_parameters(self, last_run):
+        """
+        Compute since_date/end_date and build the CybelAngel API parameters.
+        Returns (since_date, end_date, parameters) or (None, None, "sort_by=-claimed_at") if full history.
+        """
+
+        base_sort = "sort_by=claimed_at&sort_order=desc"
+
```
Fixed: the `_build_fetch_parameters` docstring is rewritten to describe the actual contract â€” it explicitly says the returned `parameters` carry `sort_by=claimed_at&sort_order=desc` (not `sort_by=-claimed_at`) and the `(None, None, base_sort)` fallback only carries the sort string when `CYBELANGEL_FETCH_PERIOD=all` (or invalid).

SamuelHassine commented on external-import/cybelangel/entrypoint.sh:
```diff
@@ -0,0 +1,7 @@
+cd /opt/opencti-connector-cybelangel
+
+# Start the connector
+python3 cybelangel.py
```
Fixed: `entrypoint.sh` now `exec`s the Python process (`exec python3 cybelangel.py`) and enables `set -e`. The shell is replaced by the Python process, so it runs as PID 1 in the container and receives `SIGTERM` / `SIGINT` directly, allowing Docker to perform a graceful shutdown.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,887 @@
+            self.helper.connector_logger.info(
+                f"Stopping processing as claimed_at {claimed_at.strftime('%Y-%m-%dT%H:%M:%SZ')} is before "
+                f"last_run value {since_date.strftime('%Y-%m-%dT%H:%M:%SZ')}"
+            )
```
Fixed: `_process_attack` now skips attacks with a missing or unparseable `claimed_at` with a warning that includes `attack_id` and the raw `claimed_at` value, instead of letting `None` flow into `Campaign.created` / `first_seen` / `last_seen` / relationship `created` (which would raise during STIX construction and abort the run). The per-claim parsing loop is also aligned with `_is_attack_too_old`, and the now-dead `if claimed_at:` branch around `campaign_date` is collapsed (`campaign_date = f" ({claimed_at.date()})"`).

SamuelHassine reviewed (dismissed):
Re-approving after pushing review-comment fixes (lowercase identity_class for canonical STIX IDs, safe claimed_at handling, robust fetch_period 'all' fallback, accurate docstring, exec for signal forwarding). All 22 review threads are resolved; locally validated with isort/black/flake8 and pylint+linter_stix_id_generator (10/10).

SamuelHassine committed:
[cybelangel] Address third-round Copilot review comments

* `cybelangel.py` (TLP:CLEAR): emit a custom STIX
  ``MarkingDefinition`` for ``TLP:CLEAR`` instead of aliasing it to
  ``stix2.TLP_WHITE``. The marking now carries
  ``x_opencti_definition_type="TLP"`` and
  ``x_opencti_definition="TLP:CLEAR"`` so OpenCTI displays the
  configured value as ``TLP:CLEAR`` (the previous alias surfaced as
  ``TLP:WHITE`` in the UI). The shape matches the canonical
  ``connectors-sdk/connectors_sdk/models/tlp_marking.py`` mapping
  (``definition_type="statement"``, ``definition={"statement":
  "custom"}``).
* `cybelangel.py` (IntrusionSet→Location / IntrusionSet→Sector
  relationships): move both relationship families out of the per-victim
  loop. Their ids are deterministic on
  ``(rel_type, source, target)`` so emitting them inside the loop
  produced N identical Relationship objects per attack (one per
  victim), inflating the bundle and risking STIX validation
  failures. The relationships are now emitted once per attack after
  the loop (mirroring the "no-victims" branch above, which already
  emitted them once). The per-victim ``IntrusionSet → victim``
  relationships stay inside the loop because they genuinely vary
  with ``identity.id``.
* `cybelangel.py` (bundle-send failure propagation):
  ``_process_attack`` now returns a ``bool`` indicating whether the
  bundle was successfully sent (``True`` is also returned for the
  deliberate "missing claimed_at" / "older than cursor" skip paths
  so a deliberately-skipped attack does not poison the run). The
  pagination loop tracks an ``all_succeeded`` flag and propagates
  it to the caller; ``opencti_bundle`` returns the same flag to
  ``process_data``, which already only advances ``last_run`` on a
  successful run. A single failed bundle no longer silently drops
  data on the next iteration.
* `src/requirements.txt`: bump ``pycti`` from ``7.260512.0`` to
  ``7.260513.0`` so the connector is aligned with the
  repository-wide pin (master and most other connectors now ship
  ``pycti==7.260513.0``).

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,914 @@
+        """
+        TLP_MAPPING = {
+            "TLP:WHITE": stix2.TLP_WHITE,
+            "TLP:CLEAR": stix2.TLP_WHITE,
```
Fixed: `TLP:CLEAR` is now emitted as a custom `stix2.MarkingDefinition` (mirroring `connectors-sdk/connectors_sdk/models/tlp_marking.py`) so the configured value is preserved end-to-end in OpenCTI:

```python
"TLP:CLEAR": stix2.MarkingDefinition(
    id=MarkingDefinition.generate_id("TLP", "TLP:CLEAR"),
    definition_type="statement",
    definition={"statement": "custom"},
    allow_custom=True,
    x_opencti_definition_type="TLP",
    x_opencti_definition="TLP:CLEAR",
),
```

Note: `MarkingDefinition.generate_id("TLP", "TLP:CLEAR")` returns the same UUID as `stix2.TLP_WHITE.id` (TLP:CLEAR is OpenCTI's modern alias for the legacy TLP:WHITE), but with the new object carrying `x_opencti_definition_type="TLP"` / `x_opencti_definition="TLP:CLEAR"`, the marking now surfaces as **"TLP:CLEAR"** in the OpenCTI UI rather than the legacy **"TLP:WHITE"** that the previous alias produced. Smoke-tested locally.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,914 @@
+                            author_org_id,
+                        )
+                    )
+                for location in locations:
+                    stix_objects.append(
+                        self._create_relationship(
+                            "targets",
+                            iset.id,
+                            location.id,
+                            claimed_at,
+                            marking_id,
+                            author_org_id,
+                        )
+                    )
```
Fixed: the `IntrusionSet ? Location` and `IntrusionSet ? Sector` relationships are no longer created inside the per-victim loop. Since `StixCoreRelationship.generate_id` is deterministic on `(rel_type, source, target)` and neither the location/sector ids nor the intrusion-set ids change per victim, emitting them inside the loop produced N identical Relationship objects per attack (one per victim) ? inflating the bundle and risking STIX validation issues.

The fix:

- inside the per-victim loop, only `IntrusionSet ? victim_identity` remains (this one genuinely varies because `identity.id` differs per victim);
- right after the loop, a new `if victim_domain_pairs:` block emits the per-attack `IntrusionSet ? Location` and `IntrusionSet ? Sector` relationships exactly once per intrusion-set (mirroring the "no-victims" branch above, which already emitted them once per attack).

No change to deterministic ids or to the produced graph shape ? just no more duplicates in the bundle.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,914 @@
+                    f"Successfully processed {len(stix_objects)} STIX objects from CybelAngel."
+                )
+            except Exception as e:
+                self.helper.connector_logger.error(f"Error creating STIX bundle: {e}")
```
Fixed: bundle-send failures now propagate end-to-end.

- `_process_attack` is now typed `-> bool`. It returns `True` on success (and on the deliberate "missing claimed_at" / "older than cursor" skip paths ? a deliberately-skipped attack must not poison the run) and `False` when `stix2.Bundle(...).serialize()` or `send_stix2_bundle()` raises. The failure is logged with the offending `attack_id` and includes the explicit hint `last_run will NOT be advanced`.
- `_fetch_and_process_pages` tracks an `all_succeeded` flag, updates it on every `_process_attack` call, and returns it on both exit paths (empty page / older-than-cursor).
- `opencti_bundle` already propagates the return value to `process_data`, and `process_data` already only sets the new `last_run` state when `success` is `True`. So a single failed bundle now leaves `last_run` untouched and the run gets retried on the next cycle. Loop semantics: we keep iterating remaining attacks after a failure (each attack is independent and re-runs are idempotent thanks to deterministic STIX ids), but the overall run is reported as failed.

SamuelHassine commented on external-import/cybelangel/src/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+pycti==7.260512.0
```
Fixed: `src/requirements.txt` now pins `pycti==7.260513.0`, matching the current master pin used by the rest of the connectors. No other dependency changes.

SamuelHassine reviewed (dismissed):
Re-approving after the third-round Copilot review. TLP:CLEAR is now emitted as a custom MarkingDefinition with x_opencti_definition='TLP:CLEAR' so the UI label is preserved (no more silent TLP:WHITE downgrade), IntrusionSet->Location/Sector relationships are emitted once per attack (no more duplicate-id Relationship objects in the bundle), bundle-send failures propagate through _process_attack -> _fetch_and_process_pages -> opencti_bundle -> process_data so last_run is no longer advanced on failure, and pycti is bumped to 7.260513.0. Validated locally with isort/black/flake8/pylint+linter_stix_id_generator (10/10) and a smoke test on the TLP map.

SamuelHassine committed:
[cybelangel] Address fourth-round Copilot review comments

* `src/cybelangel.py`: the `__main__` last-resort exception handler now
  exits with status code `1` instead of `0`. Returning `0` after a crash
  masked the failure for container supervisors, CI runners and Docker
  restart policies (they only restart / fail-build on non-zero exits).
  A short `time.sleep(10)` is preserved so the operator has time to see
  the error in the logs before the supervisor restarts the container.
* `README.md`: the manual-deployment instructions now point at the
  correct file paths. The sample lives at `src/config.yml.sample` and
  the loader in `cybelangel.py` reads `src/config.yml` (the file next
  to `cybelangel.py`), so the README now spells out
  `cp src/config.yml.sample src/config.yml` and clarifies that the
  config file must sit next to the entry point.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,966 @@
+    except Exception as e:
+        print(f"Error running CybelAngel connector: {e}")
+        time.sleep(10)
+        sys.exit(0)
```
Fixed by e9efac6e2. The `__main__` last-resort exception handler now exits with `sys.exit(1)` (with an explanatory comment), so container supervisors / CI runners / Docker `restart: unless-stopped` policies no longer mistake a crash for a clean shutdown:

```python
if __name__ == "__main__":
    try:
        cybelAngelConnector = CybelAngel()
        cybelAngelConnector.run()
    except Exception as e:
        # Non-zero exit so container supervisors / CI / restart policies
        # do not mistake a crash for a successful run.
        print(f"Error running CybelAngel connector: {e}")
        time.sleep(10)
        sys.exit(1)
```

The short `time.sleep(10)` is kept on purpose: when several connectors crash together (e.g. a temporary platform outage), it spreads the restart attempts over a small window so the OpenCTI platform doesn't get hammered by simultaneous re-auth attempts the moment it comes back up. Removing the sleep would not change the exit semantics but would make crash loops noisier in practice.

SamuelHassine commented on external-import/cybelangel/README.md:
````diff
@@ -0,0 +1,99 @@
+Create `config.yml` from `config.yml.sample`, install dependencies and
+start the connector:
+
+```shell
````
Fixed by e9efac6e2. The manual-deployment instructions now point at the actual file paths the connector loads from (`src/config.yml`, alongside `cybelangel.py`):

```markdown
### Manual

The connector loads its configuration from `src/config.yml` (same directory
as `cybelangel.py`). Create that file from the provided sample, install
dependencies and start the connector:

\`\`\`shell
cp src/config.yml.sample src/config.yml
# edit src/config.yml and fill in opencti.token, cybelangel.client_id,
# cybelangel.client_secret, ...
pip3 install -r src/requirements.txt
python3 src/cybelangel.py
\`\`\`
```

So an operator following the README copy-pastes the right file (`src/config.yml.sample` -> `src/config.yml`) and the loader (`os.path.dirname(os.path.abspath(__file__)) + "/config.yml"`) finds it on the first try.

SamuelHassine reviewed (dismissed):
Re-approving after the fourth-round Copilot review fixes, head commit e9efac6e2.

Both remaining Copilot threads on the previous commit are addressed and resolved:

- `src/cybelangel.py`: the `__main__` last-resort exception handler now exits with `sys.exit(1)` so container supervisors / CI runners / Docker restart policies do not mistake a crash for a clean run. The short `time.sleep(10)` is kept on purpose to spread restart attempts when several connectors crash together (e.g. during a temporary OpenCTI outage).
- `README.md`: the manual-deployment section now spells out the actual file paths the loader uses (`cp src/config.yml.sample src/config.yml`, alongside `cybelangel.py`), matching the `os.path.dirname(os.path.abspath(__file__)) + "/config.yml"` lookup in the connector.

CI status on e9efac6e2:

- GitHub Actions: PR title convention, signed commits, baseline coverage, issue-link, do-not-merge, test detection, codecov — all green.
- CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` — all green.
- `filigran/cla` — already signed (organization member).

`black --check`, `isort --profile black --check`, `flake8 --ignore=E,W` and the custom `linter_stix_id_generator` pylint plugin (10/10) all report a clean run locally on the connector source. Branch is `MERGEABLE` (no conflict with `master`).

SamuelHassine committed:
[cybelangel] Address fifth-round Copilot review comments

- OAuth2 `audience` is no longer hard-coded to `https://platform.cybelangel.com/`;
  it is now a new optional config variable (`CYBELANGEL_AUDIENCE` /
  `cybelangel.audience`) that defaults to the configured
  `cybelangel.api_url` with a trailing slash. This keeps the audience in
  sync with the API host when operators point the connector at a
  different CybelAngel environment.
- `authenticate()` no longer logs `Retrying in N seconds...` on the
  final attempt. The terminal failure now emits a dedicated
  `Authentication failed after N attempts, giving up.` error message
  and the retry log only fires when another attempt will actually
  follow.
- Default `CONNECTOR_LOG_LEVEL` in the connector is now `info` to match
  the value documented in `README.md` (and to align with the connector
  template README).
- Dropped the unused `libxslt` / `libxml2` (and their `-dev` variants)
  from the `Dockerfile`. The connector's Python dependencies (`pycti`,
  `requests`, `stix2`, `PyYAML`) do not need `lxml` at runtime,
  shrinking the final image and reducing the attack surface.
- Renamed the `__main__` local from camelCase `cybelAngelConnector` to
  snake_case `connector` for consistency with the rest of the repo.

`README.md`, `docker-compose.yml` and `config.yml.sample` updated to
document the new optional `CYBELANGEL_AUDIENCE` variable.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,968 @@
+            :param max_retries: Maximum number of retry attempts.
+        """
+
+        auth_data = {
+            "client_id": self.cybelangel_client_id,
+            "client_secret": self.cybelangel_client_secret,
+            "audience": "https://platform.cybelangel.com/",
+            "grant_type": "client_credentials",
+        }
```
Good catch — fixed in d7f36f14.

The `audience` is no longer hard-coded. It now comes from a new optional config variable `CYBELANGEL_AUDIENCE` (`cybelangel.audience`) that defaults to `self.cybelangel_api_url.rstrip("/") + "/"`, so the audience stays consistent with the configured API host out of the box and can still be overridden explicitly when an environment needs a different value.

See `src/cybelangel.py` (config block + `authenticate()`) and the new row in `README.md` / `docker-compose.yml` / `config.yml.sample`.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,968 @@
+                    f"Retrying in {delay} seconds... (Attempt {attempt}/{max_retries})"
+                )
+                if attempt < max_retries:
+                    time.sleep(delay)
```
Fixed in d7f36f14.

`authenticate()` no longer logs `Retrying in N seconds...` on the terminal attempt. The retry log is now gated on `attempt < max_retries` (matching the existing `time.sleep` guard) and a dedicated `Authentication failed after N attempts, giving up.` error is emitted when we run out of attempts, so the logs make the actual outcome unambiguous.

SamuelHassine commented on external-import/cybelangel/Dockerfile:
```diff
@@ -0,0 +1,17 @@
+# hadolint ignore=DL3003
+RUN apk --no-cache add git build-base libmagic libxslt libxslt-dev libxml2 libxml2-dev && \
+    pip3 install --no-cache-dir -r requirements.txt && \
+    apk del git build-base libxslt-dev libxml2-dev
```
Agreed — fixed in d7f36f14.

`libxslt`, `libxslt-dev`, `libxml2` and `libxml2-dev` are gone. The runtime deps (`pycti`, `requests`, `stix2`, `PyYAML`) don't need `lxml`, so the `RUN` line is now just `git build-base libmagic` (+ `apk del git build-base` at the end), which shrinks the final image and the attack surface.

SamuelHassine commented on external-import/cybelangel/README.md:
```diff
@@ -0,0 +1,103 @@
+| Connector Type     | `type`              | `CONNECTOR_TYPE`            | `EXTERNAL_IMPORT` | No    | Must be `EXTERNAL_IMPORT`.                                                 |
+| Connector Name     | `name`              | `CONNECTOR_NAME`            | `CybelAngel`  | No        | Name of the connector as displayed in the OpenCTI UI.                      |
+| Connector Scope    | `scope`             | `CONNECTOR_SCOPE`           | `all`         | No        | Scope of the connector.                                                    |
+| Log Level          | `log_level`         | `CONNECTOR_LOG_LEVEL`       | `info`        | No        | One of `debug`, `info`, `warn`, `error`.                                   |
```
Fixed in d7f36f14.

Aligned the code default with the documented one — `CONNECTOR_LOG_LEVEL` now defaults to `info` (matching what `README.md` advertises and what the connector template README documents). Operators relying on defaults will now get the expected verbosity.

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,968 @@
+if __name__ == "__main__":
+    try:
+        cybelAngelConnector = CybelAngel()
+        cybelAngelConnector.run()
```
Fixed in d7f36f14.

Renamed the `__main__` local from camelCase `cybelAngelConnector` to snake_case `connector`, matching the convention used elsewhere in the repo.

SamuelHassine reviewed (approved):
Re-approving after the fifth-round Copilot review fixes, head commit `d7f36f14`.

All 5 remaining Copilot threads on the previous head (`e9efac6e2`) are now addressed and resolved:

- `src/cybelangel.py`: OAuth2 `audience` is no longer hard-coded. It now comes from a new optional config variable `CYBELANGEL_AUDIENCE` (`cybelangel.audience`) defaulting to `self.cybelangel_api_url.rstrip("/") + "/"`, so it stays consistent with the configured API host when an operator points the connector at a different CybelAngel environment.
- `src/cybelangel.py`: `authenticate()` no longer logs `Retrying in N seconds...` on the final attempt. The retry log is gated on `attempt < max_retries` and a dedicated `Authentication failed after N attempts, giving up.` error is emitted on terminal failure.
- `src/cybelangel.py`: Default `CONNECTOR_LOG_LEVEL` is now `info`, matching what `README.md` advertises and what the connector template README documents.
- `src/cybelangel.py`: Renamed the `__main__` local from camelCase `cybelAngelConnector` to snake_case `connector` for consistency with the rest of the repo.
- `Dockerfile`: Dropped `libxslt`/`libxml2` (and their `-dev` variants). The connector's runtime deps (`pycti`, `requests`, `stix2`, `PyYAML`) don't need `lxml`, so the final image is leaner.
- `README.md` / `docker-compose.yml` / `config.yml.sample` updated to document the new optional `CYBELANGEL_AUDIENCE` variable.

CI status on `d7f36f14`:

- GitHub Actions: PR title convention, signed commits, baseline coverage, issue-link, do-not-merge, test detection, codecov — all green.
- CircleCI: `ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest` — all green.
- `filigran/cla` — signed.

`black --check`, `isort --profile black --check`, `flake8 --ignore=E,W` and the custom `linter_stix_id_generator` pylint plugin (10/10) all report a clean run locally on the connector source. Branch is `MERGEABLE` (no conflict with `master`).

SamuelHassine commented on external-import/cybelangel/src/cybelangel.py:
```diff
@@ -0,0 +1,989 @@
+                # case which means "no date filter".
+                self.helper.connector_logger.warning(
+                    "Invalid last_run format, falling back to CYBELANGEL_FETCH_PERIOD",
+                    {"last_run": last_run},
```
Not changed — this is a false positive specific to this comment.

`self.helper.connector_logger` is not a stdlib `logging.Logger`, it is pycti's `AppLogger` wrapper (`pycti.utils.opencti_logger.AppLogger`) whose signature is:

```python
def warning(self, message, meta=None):
    self.local_logger.warning(message, extra=AppLogger.prepare_meta(meta))
```

The second positional argument is `meta` (a `dict`), not `*args`. `meta` is forwarded to the stdlib logger via the `extra=` keyword, **not** as `%`-style formatting arguments, so it cannot trigger `not all arguments converted` errors and the structured context is preserved.

This is the same positional-dict idiom pycti itself uses internally, e.g. in `opencti_connector_helper.py`:

```python
self.helper.connector_logger.info("Message ack", {"tag": method.delivery_tag})
self.connector_logger.error("Error pinging the API", {"reason": str(e)})
```

So lines 380 / 407 / 583 (and the two suppressed sites at 410 / 586) are already using the supported structured-logging style for `AppLogger`. No code change needed.

SamuelHassine reviewed (approved):
Final review pass on head commit `d7f36f14`.

The single remaining Copilot review thread (line 380 / 407 / 583 — "`connector_logger.warning()` is called with a dict as a second positional argument") has been replied to and resolved. It is a false positive: `self.helper.connector_logger` is pycti's `AppLogger`, not a stdlib `logging.Logger`. Its signature is `warning(message, meta=None)` and `meta` is forwarded to the stdlib logger via the `extra=` keyword, **not** as `%`-style formatting arguments — so a positional dict cannot trigger `not all arguments converted` errors and the structured context is preserved. This is the exact same idiom pycti itself uses internally (`opencti_connector_helper.py:558` and `:1035`), so no code change is needed.

Status:

- All review threads (Copilot + maintainer) are now resolved.
- CI on `d7f36f14`: GitHub Actions and CircleCI fully green (`ensure_formatting`, `base_linter`, `linter`, `test`, `build_manifest`, `codecov/patch`, `codecov/project`, signed commits, PR title convention, linked issue, do-not-merge, baseline coverage). `filigran/cla` already signed.
- `black --check`, `isort --profile black --check`, `flake8 --ignore=E,W` and the custom `linter_stix_id_generator` pylint plugin (10/10) all clean locally on the connector source.
- Branch state: `MERGEABLE` / `mergeStateStatus: CLEAN` against `master` — no conflict.
- PR is not a draft, title `[cybelangel] New external-import connector for the CybelAngel platform` matches the `^\[\w+.+\].+$` convention enforced by `pr-check-conventions.yml`, the linked issue (`#4450`) is set, and labels `filigran team` / `connector: cybelangel` are applied.

Ready to merge.
